lib/compat/wordpress-6.1/rest-api.php

```diff
@@ -93,3 +93,84 @@ function gutenberg_register_has_archive_on_post_types_endpoint() {
 	);
 }
 add_action( 'rest_api_init', 'gutenberg_register_has_archive_on_post_types_endpoint' );
+
+/**
+ * Use custom REST API Controller for Navigation Posts
+ *
+ * @param array $args the post type arguments.
+ * @return array the filtered post type arguments with the new REST Controller set.
+ */
+function gutenberg_update_navigation_rest_controller( $args, $post_type ) {
+	if ( in_array( $post_type, array( 'wp_navigation' ), true ) ) {
+		// Original set in
+		// https://github.com/WordPress/wordpress-develop/blob/6cbed78c94b9d8c6a9b4c8b472b88ee0cd56528c/src/wp-includes/post.php#L528.
+		$args['rest_controller_class'] = 'Gutenberg_REST_Navigation_Controller';
+	}
+	return $args;
+}
+add_filter( 'register_post_type_args', 'gutenberg_update_navigation_rest_controller', 10, 2 );
+
+function gutenberg_transform_slug_to_post_id( $response, $handler, WP_REST_Request $request ) {
+
+	// Create (POST) requests send the `slug` as part of the request **body** for the Post to be created.
+	// As a result the `slug` in POST requests should not be transformed into an ID.
+	$should_preserve_slug = 'POST' === $request->get_method();
```

Comment:
Sorry, I'm not following this completely. Could you explain a bit more as to why this is needed?

Checking that the method is `POST` doesn't mean that this is a creation request, you can `POST` to a single item to update it, ie `/wp/v2/navigation/5` works. We'd need to check that the route is exactly `/wp/v2/navigation` and `POST`.

Reply:
Yeh this was a mess 😄 

I misread the problem. Basically:

- a `slug` param can exist both as a _body_ param and a _url_ param.
- if you want to update a Navigation post you must send the `slug` of that post in the URL (eg: `/wp/navigation/my-slug`) in order that we can map the slug to a Post and set the Post ID.
- however if you want to update the `slug` of the Navigation then you end up with x2 `slug` params - one in the URL (as above) and one in the POST body. These will be different because one is acting as an "ID" whereas the other is new data you'd like updated.

I've now updated the implementation to always look for `slug` in the url param. This ensures that the `slug` in the body is never used to look up the post.

I've left comments...